### PR TITLE
VCST-4772 - AutoMapper Removing

### PR DESCRIPTION
**Pull request by DmitryGrishinVirtoworks**
## Description
Breaking change:
Public constructors of CreateReviewCommandHandler, EvalVendorRatingMiddleware, EvalProductRatingMiddleware and EvalProductVendorRatingMiddleware changed (removed AutoMapper.IMapper, added ICustomerReviewMapper). The public RatingProfile AutoMapper profile was removed — any external code calling _mapper.Map<ExpRating>(...), _mapper.Map<ExpVendorRating>(...) or _mapper.Map<CustomerReview>(command) directly against the host-wide IMapper will still compile but will throw AutoMapperMappingException at runtime after this package is updated, since those type maps no longer exist. Please version this as a breaking change.

Security status (GHSA-rvv3-g6hj-g44x):
This PR removes this module's own usage of AutoMapper, but the AutoMapper 12.0.1 package itself remains in the resolved dependency graph transitively via VirtoCommerce.Xapi.Core, whose AddSchema() extension still calls services.AddAutoMapper(...). The DoS advisory and AutoMapper 15+ licensing exposure are only fully closed once vc-module-x-api drops its own AutoMapper dependency and this module bumps that package reference. This PR is a necessary step but should not be treated as closing the security ticket on its own.

### Jira-link:
https://virtocommerce.atlassian.net/browse/VCST-4772
### Artifact URL:
https://vc3prerelease.blob.core.windows.net/packages/VirtoCommerce.CustomerReviews_3.1006.0-pr-82-3d7b.zip

### CODE REVIEW

File: src/VirtoCommerce.CustomerReviews.ExperienceApi/Extensions/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -37,6 +38,11 @@ public static IServiceCollection AddExperienceApi(this IServiceCollection servic
         });
 
         serviceCollection.AddTransient<ReviewValidator>();
+
+        serviceCollection.AddTransient<IRatingEntityDtoMapper, RatingEntityDtoMapper>();
```
Comment (alexeyshibanov) [inline src/VirtoCommerce.CustomerReviews.ExperienceApi/Extensions/ServiceCollectionExtensions.cs]: These mappers are stateless (no fields, no ctor dependencies), so `AddSingleton` is the correct lifetime — `AddTransient` just allocates per resolve. With the single `ICustomerReviewMapper` facade this collapses to one registration:

```csharp
serviceCollection.AddSingleton<ICustomerReviewMapper, CustomerReviewMapper>();
```

File: src/VirtoCommerce.CustomerReviews.ExperienceApi/Mapping/RatingEntityDtoMapper.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using VirtoCommerce.Xapi.Core.Models;
+
+namespace VirtoCommerce.CustomerReviews.ExperienceApi.Mapping;
+
+public class RatingEntityDtoMapper : IRatingEntityDtoMapper
+{
+    public ExpRating ToExpRating(RatingEntityDto source)
+    {
+        var result = AbstractTypeFactory<ExpRating>.TryCreateInstance();
+        result.Value = source.Value;
```
Comment (alexeyshibanov) [inline src/VirtoCommerce.CustomerReviews.ExperienceApi/Mapping/RatingEntityDtoMapper.cs]: Null-semantics regression vs AutoMapper: `IMapper.Map<T>((object)null)` returned `null`, the hand-written mapper now throws NRE on a null `source`. Current call sites guard via `TryGetValue`, so it's latent, but as the series template this deserves an explicit decision: either `if (source is null) return null;` at the top of each `To*` method, or a stated "callers must not pass null" contract. Same applies to the other two mappers.

File: src/VirtoCommerce.CustomerReviews.ExperienceApi/Commands/CreateReviewCommandHandler.cs [hunk 3]
```diff
@@ -24,28 +24,28 @@ public class CreateReviewCommandHandler : IRequestHandler<CreateReviewCommand, C
...
 
     public CreateReviewCommandHandler(
         ICustomerReviewService reviewService,
-        IMapper mapper,
         IMemberService memberService,
         IFileUploadService fileUploadService,
         Func<UserManager<ApplicationUser>> userManagerFactory,
-        ReviewValidator reviewValidator)
+        ReviewValidator reviewValidator,
+        ICreateReviewCommandMapper mapper)
```
Comment (alexeyshibanov) [inline src/VirtoCommerce.CustomerReviews.ExperienceApi/Commands/CreateReviewCommandHandler.cs]: Nit: the new dependency was appended last instead of replacing `IMapper` in place, so the field block and parameter list are reordered beyond what the change required (`_mapper` is now detached from the other injected fields). In-place replacement keeps the diff minimal — the ctor is a compile-time break for downstream subclasses either way, but there's no need to also reorder.

File: tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        {
+            var source = new CreateReviewCommand
+            {
+                StoreId = "Store1",
+                EntityId = "P1",
+                EntityType = "Product",
+                UserId = "U1",
+                Review = "Great product",
+                Rating = 5,
+                ImageUrls = ["/api/files/img1"],
```
Comment (alexeyshibanov) [inline tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs:94]: The test sets `ImageUrls` but never asserts what happens to it. The non-mapping is intentional (the old profile didn't map it either — the handler populates `review.Images` separately), so pin the intent with `Assert.Null(result.Images);` — otherwise a future reader can't tell omission from oversight.

While here, cheap edge cases worth adding for the series template: null `source` into each mapper (documents the null contract), and a test that resolves mappers from DI rather than `new`-ing them, so the `ServiceCollectionExtensions` registrations are covered too.

**Review by alexeyshibanov — CHANGES_REQUESTED** [review]
## Direction: right. Shape: let's set the series canon here

This PR is the first of the VCST-4772 series (~10 active modules use AutoMapper), so whatever shape lands here will be replicated. The migration itself is well executed: field-by-field fidelity with the deleted profile is exact, `AbstractTypeFactory` construction is preserved — and for `CustomerReview` actually *fixed*: the old plain `CreateMap<CreateReviewCommand, CustomerReview>()` constructed the destination via `new`, ignoring registered type overrides, while the new mapper honors them (worth a release-note line on its own). Tests assert real values. Requesting changes on the mapper **shape** and compat signalling, not on the direction.

### 1. One module mapper facade instead of interface-per-mapping

Three single-method public interfaces for three ~4-line mappings means 3 DI registrations, 3 ctor dependencies, and 3 permanent compat obligations where one suffices — and as the series template it multiplies across ~10 modules. Proposal:

```csharp
public interface ICustomerReviewMapper
{
    ExpRating ToExpRating(RatingEntityDto source);
    ExpVendorRating ToExpVendorRating(RatingEntityStoreDto source);
    CustomerReview ToCustomerReview(CreateReviewCommand source);
}

public class CustomerReviewMapper : ICustomerReviewMapper
{
    public virtual ExpRating ToExpRating(RatingEntityDto source) { /* unchanged body */ }
    // ...
}
```

- `I<Module>Mapper` keeps the AutoMapper lineage visible at the type level (one mapping surface per module, like the old profile was).
- `public virtual` methods let a downstream override ONE mapping and keep `base` for the rest — cheaper than re-implementing an interface, and consistent with how VC services are DI-substituted downstream.
- Method naming `To<Dest>(source)`: bare `Map(source)` overloads cannot survive the series — C# cannot overload by return type, and sibling modules have same-source/multi-destination maps (`SearchDocument` -> `ExpProduct`/`ExpCategory` in x-catalog; `IList<IFilter>` -> 6 different destinations across the x-api/x-catalog facet profiles). `To<Dest>` is collision-proof and matches the existing `ToModel` vocabulary.

### 2. Series canon for the harder modules (forward-looking, so later PRs don't diverge)

- **In-place mapping** (`mapper.Map(source, target)` sites, e.g. profile-experience-api `UpdateContactCommand -> Contact`): `void ApplyTo(TSource source, TTarget target)` — same argument order as AutoMapper's `Map(source, destination)`; `ForMember(..., Condition(...))` becomes an explicit `if`, `Ignore()` becomes absence of the line, both visible at the mapping site instead of hidden in a profile.
- **Caller context**: the x-api / x-catalog / x-cart / x-order profiles read `context.Items["cultureName" | "currency" | "pricelists" | ...]`. The replacement should be a **typed** `MappingContext` (well-known properties, module-specific subclasses), not a string-keyed bag — hand-written overloads have no fixed `Map<T>` signature, so the bag is no longer forced. Natural home: `Xapi.Core`.
- **No generic `IMapper<TSource, TDest>` contract in x-api**: no generic mapping call sites exist in the active modules today; a shared contract would couple all modules' evolution to one interface for no current consumer. Easy to add later if a real generic consumer appears.
- **Convention-debt guard**: AutoMapper silently picked up newly added same-named properties; hand-written mappers lose that. A small reflection-based test utility (assert every readable/writable same-named property pair is either copied or in an explicit ignore list) restores the guarantee — can ship once next to `MappingContext` and be reused by every module's mapping tests.

### 3. Breaking-change signalling (needed regardless of shape)

- Deleting the public `RatingProfile` removes its three maps from the **host-wide** `IMapper` (they were picked up by x-api's `AddSchema` -> `services.AddAutoMapper(assemblyMarker)` scan). Downstream code calling `_mapper.Map<ExpRating>(...)` / `Map<ExpVendorRating>(...)` / `Map<CustomerReview>(command)` keeps compiling and fails at runtime with `AutoMapperMappingException` after the package bump; the public ctors of `CreateReviewCommandHandler` and the three middlewares changed as well. Please flag this in release notes and version accordingly.
- The AutoMapper 12.0.1 package itself **stays** in this module's resolved dependency graph (transitive via `VirtoCommerce.Xapi.Core`), and `AddSchema` still calls `AddAutoMapper`. This PR is the necessary consumer-side step, but the GHSA-rvv3-g6hj-g44x exposure only ends once x-api drops AutoMapper and this module bumps that reference — worth stating in the PR description so the ticket isn't considered closed by module-level PRs alone.

File: src/VirtoCommerce.CustomerReviews.ExperienceApi/Extensions/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (DmitryGrishinVirtoworks) [reply src/VirtoCommerce.CustomerReviews.ExperienceApi/Extensions/ServiceCollectionExtensions.cs]: Done

File: src/VirtoCommerce.CustomerReviews.ExperienceApi/Mapping/RatingEntityDtoMapper.cs (hunk 2 above)
Comment (DmitryGrishinVirtoworks) [reply src/VirtoCommerce.CustomerReviews.ExperienceApi/Mapping/RatingEntityDtoMapper.cs]: Done

File: src/VirtoCommerce.CustomerReviews.ExperienceApi/Commands/CreateReviewCommandHandler.cs (hunk 3 above)
Comment (DmitryGrishinVirtoworks) [reply src/VirtoCommerce.CustomerReviews.ExperienceApi/Commands/CreateReviewCommandHandler.cs]: Done

File: tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            var result = new CustomerReviewMapper().ToCustomerReview(null);
+
+            Assert.Null(result);
+        }
+
+        [Fact]
+        public void CustomerReviewMapper_Is_Resolvable_From_DI()
+        {
+            var services = new ServiceCollection();
+            services.AddSingleton<ICustomerReviewMapper, CustomerReviewMapper>();
```
Comment (alexeyshibanov) [inline tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs]: This test builds its own `ServiceCollection` and calls `AddSingleton<ICustomerReviewMapper, CustomerReviewMapper>()` itself — a copy of the production line. Delete the registration from `ServiceCollectionExtensions.cs` and this test still passes green, so what it actually asserts is that `Microsoft.Extensions.DependencyInjection` resolves a registration, not that the module makes one.

Quick way to check any test of this shape: mentally remove the production line and ask whether the test goes red. If it doesn't, it is measuring something other than what its name claims.

Two honest options — either resolve from the real registration:

```csharp
var services = new ServiceCollection();
services.AddExperienceApi();
// assert ICustomerReviewMapper resolves
```

(worth checking whether `AddExperienceApi` runs in a unit test — it constructs a `GraphQLBuilder` and calls `AddSchema`, so it may need more infrastructure than a bare `ServiceCollection`), or simply drop the test: the facade has no resolution mechanism beyond the single registration line, so there is nothing left to guard once the copy is removed. Non-blocking either way, but worth settling before the pattern is copied into the other eight modules of the series.

**Review by alexeyshibanov — DISMISSED** [review]
Approving — the canonical shape landed, and this PR is now a good template for the rest of the VCST-4772 series. Verified locally at `bde08d97`: build clean (0 warnings), unit tests green (the one failing test is the pre-existing `CustomerReviewsTests.CanDoCrudAndSearch`, which resolves `TestData` through a path with Windows separators and fails on Linux — untouched by this PR).

What landed well:

- Six types collapsed into one `ICustomerReviewMapper` / `CustomerReviewMapper` facade with `public virtual` methods and a single `AddSingleton` registration — a downstream can now override one mapping and keep `base` for the rest, which the previous interface-per-mapping shape did not allow cheaply.
- Null semantics restored on all three methods, matching what `IMapper.Map<T>(null)` did before. `== null` is the right spelling here — it matches this repo's existing idiom.
- Constructor parameter positions are back to what `dev` had, in the handler **and** in all three middlewares (my note only pointed at the handler; the middlewares had drifted too). The diff against `dev` is now a plain type swap.
- The `Assert.Null(result.Images)` comment names the actual mechanism (`SaveImages` populating `review.Images` after resolving uploads), so a future "tidy-up" that adds an `ImageUrls` mapping — which would double-handle the files — now has something to stop it.

Two non-blocking items, neither of which needs to hold the merge:

**1. PR description (highest value of the two).** The `## Description` section is still empty, and this change has downstream-visible effects worth recording there or in release notes:

- Removing the public `RatingProfile` drops three maps from the **host-wide** `IMapper` (they were merged in via x-api's `AddSchema` -> `AddAutoMapper` assembly scan). Downstream code calling `_mapper.Map<ExpRating>(...)`, `Map<ExpVendorRating>(...)` or `Map<CustomerReview>(command)` keeps compiling and starts throwing `AutoMapperMappingException` after the package bump.
- Public constructors changed on `CreateReviewCommandHandler` and the three middlewares.
- AutoMapper 12.0.1 still resolves into this module's dependency graph transitively via `VirtoCommerce.Xapi.Core`, and `AddSchema` still calls `AddAutoMapper` (now finding no profiles here). So GHSA-rvv3-g6hj-g44x is not yet closed for this module — this PR is a prerequisite, and the exposure ends when x-api drops AutoMapper and this module bumps that reference. Stating it in the description keeps the ticket from looking done once the module-level PRs merge.

**2. `CustomerReviewMapper_Is_Resolvable_From_DI` does not cover the registration it is named after** — see the inline note.

For the rest of the series (not this PR): in-place mapping sites (`mapper.Map(source, target)`, e.g. profile-experience-api's `UpdateContactCommand` -> `Contact`) map to `ApplyTo(source, target)`, and the modules whose profiles read `context.Items` (x-api, x-catalog, x-cart, x-order) need a typed `MappingContext` passed as a parameter rather than a string-keyed bag — keeping mappers stateless singletons with empty constructors. A shared reflection-based guard test (every same-named readable/writable property pair is either copied or explicitly ignored) would restore the property-sync guarantee AutoMapper's conventions provided; `ToCustomerReview` here is a legitimate consumer of it once that utility exists.

File: tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs (hunk 5 above)
Comment (DmitryGrishinVirtoworks) [reply tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs]: Done

File: tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+        [Fact]
+        public void AddExperienceApi_Registers_CustomerReviewMapper()
+        {
+            var services = new ServiceCollection();
+            services.AddExperienceApi();
+
+            var mapper = services.BuildServiceProvider().GetRequiredService<ICustomerReviewMapper>();
+
+            Assert.IsType<CustomerReviewMapper>(mapper);
```
Comment (alexeyshibanov) [inline tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs]: Optional (not blocking this PR), but worth settling here since the other eight modules will copy this test: asserting on the `ServiceCollection` descriptor instead of a built provider is strictly stronger, and it also drops the undisposed `ServiceProvider`.

```suggestion
            var descriptor = services.Single(x => x.ServiceType == typeof(ICustomerReviewMapper));

            Assert.Equal(typeof(CustomerReviewMapper), descriptor.ImplementationType);
            Assert.Equal(ServiceLifetime.Singleton, descriptor.Lifetime);
```

Why it is stronger: the current form passes for **any** lifetime, because the service resolves either way. I ran both variants locally against this branch —

- registration commented out: current form red (`No service for type ... registered`), suggested form red (`Sequence contains no matching element`) — both catch this;
- `AddSingleton` changed to `AddTransient`: current form stays **green**, suggested form goes red (`Expected: Singleton, Actual: Transient`).

Since "stateless mapper, empty constructor, singleton" is the shape the series standardises on, the lifetime is part of what the test should pin — otherwise a later `AddTransient` slips through silently. Secondary benefit: no `BuildServiceProvider()` means there is no `IDisposable` provider left undisposed, so the test class does not need to become `IDisposable` to be strictly correct.

Trade-off, for completeness: the LINQ failure message (`Sequence contains no matching element`) is less self-explanatory than `GetRequiredService`'s. If that matters, `SingleOrDefault(...)` plus an `Assert.NotNull` with an explicit message reads better on failure. `using System.Linq;` and `Microsoft.Extensions.DependencyInjection` are already imported here, so the snippet compiles as-is (verified).

**Review by alexeyshibanov — COMMENTED** [review]
The test now genuinely covers the registration — verified by mutation: commenting out `AddSingleton<ICustomerReviewMapper, CustomerReviewMapper>()` in `ServiceCollectionExtensions.cs` turns it red (`No service for type 'ICustomerReviewMapper' has been registered`). Thanks for switching to the real `AddExperienceApi()`. The PR description covers the breaking change and the transitive-AutoMapper status well too.

One optional refinement left, inline — relevant mostly because this file is the template the rest of the VCST-4772 series will copy.

File: tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs (hunk 6 above)
Comment (DmitryGrishinVirtoworks) [reply tests/VirtoCommerce.CustomerReviews.Test/RatingMappingTests.cs]: Done

**Review by alexeyshibanov — APPROVED** [review]
Approving `3d7bdf6d` — everything raised earlier is closed, and the test came out better than either variant I proposed: the descriptor assertions keep the lifetime pinned, and `SingleOrDefault` + the explicit message keeps the failure readable.

Verified locally against this commit, not just read: build clean (0 warnings, re-checked with `--no-incremental`; no xUnit-analyzer or nullable warning on the `Assert.True(descriptor != null, ...)` → `descriptor.ImplementationType` pattern), unit tests green apart from the pre-existing `CustomerReviewsTests.CanDoCrudAndSearch`, which resolves `TestData` through a path with Windows separators and cannot pass on Linux — untouched by this PR and worth a separate fix. Both mutations of the production registration produce distinct, self-explanatory failures:

- registration removed → `ICustomerReviewMapper is not registered in AddExperienceApi()`
- `AddSingleton` → `AddTransient` → `Assert.Equal() Failure: Expected: Singleton, Actual: Transient`

That second one is what makes this test worth copying: the lifetime is part of the shape the series standardises on (stateless mapper, empty constructor, singleton), and nothing else in the build would catch a silent downgrade to `AddTransient`.

Summary of what this PR establishes as the template for the remaining VCST-4772 modules: one `I<Module>Mapper` facade with `public virtual To<Destination>` methods, a single `AddSingleton` registration, null-in/null-out preserved, and a registration test asserting both implementation type and lifetime. For the modules whose profiles read `context.Items` (x-api, x-catalog, x-cart, x-order) the missing pieces are a typed `MappingContext` passed as a parameter — keeping mappers stateless with empty constructors — and `ApplyTo(source, target)` for the in-place mapping sites (profile-experience-api's `UpdateContactCommand` → `Contact`), where a `ForMember` `Condition` becomes an explicit `if` and an `Ignore` becomes an absent line.

The PR description already records the breaking change and states that AutoMapper stays in the graph transitively via `Xapi.Core`, so the security ticket is not closed by this module alone — thanks for spelling both out.
